### PR TITLE
[inductor] Skip the sub-process pool until it's ready

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133510
* __->__ #133508

Summary: Torch-compiling a quick script can be a bit slower than it needs to be: even though we initialize the subprocess pool early, it still might not be ready by the time we try to compile the first Triton kernel. Instead, let's use the single-threaded path until the pool has successfully completed a no-op job.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by masnesral** [comment]
Compile time looks fine: https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2009%20Aug%202024%2013%3A48%3A23%20GMT&stopTime=Fri%2C%2016%20Aug%202024%2013%3A48%3A23%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/masnesral/105/head&lCommit=45ae57f49e438855c9411d58524edd198958321b&rBranch=main&rCommit=a6413d2924be7df5c1afc3bb0ad09fdbb7519101

**Comment by Chillee** [comment]
Any benchmarks on how this affects a simple script?

**Comment by masnesral** [comment]
@Chillee 

> Any benchmarks on how this affects a simple script?

Ooops! I meant to type the following...

Following up on your DM, I used your example:
```
import torch
torch.set_default_device('cuda')

GLOBAL = True
@torch.compile
def f(x):
    if GLOBAL:
        return x.cos().cos()
    else:
        return x.sin().sin()

x = torch.randn(5)
f(x)
```

I actually couldn't repro the perf difference you reported (17 seconds to 11). I'm hoping that what you really saw was a cache hit on the 2nd run with you tried compile_threads=1. On my devserver (with no caching), it's more like 7s vs. 5.5s.

When I play around with various random examples and make sure the subprocess pool is warm before compile, I only see a max of 100ms extra latency due to the subprocess overhead. But generally the subprocess approach is much better overall because we can overlap compilation (when there is compilation to overlap). So I think the single-threaded compile is practically only faster if the subprocess pool hasn't completely warmed up yet. 

Long story short, the hack in this PR removes the 1.5s overhead I see and I get 5.5s with and without compile_threads=1

**Comment by masnesral** [comment]
@Chillee , see my last update. WDYT? I don't think this change is that important and it doesn't make a big difference in compile-time, but you had mentioned being annoyed at the overhead of quick scripts.

**Comment by Chillee** [comment]
@masnesral It seems to make a decent difference for me.

It goes from 10 seconds to 7 seconds (without compilation at all it's 3 seconds).

image

**Comment by masnesral** [comment]
@Chillee, I can't tell the state of your caches from that screenshot though. Is it possible the 2nd run is benefiting from an FX graph cache hit and that's the bulk of the difference? What do you see with `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1` for both runs?

**Comment by Chillee** [comment]
@masnesral No, both runs are with warm caches.

**Comment by masnesral** [comment]
@Chillee, ok cool. Do you want to accept this PR then ;) ?

**Review by Chillee — APPROVED** [review]
